Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/TestCertificateRepository.kt

```diff
@@ -0,0 +1,329 @@
+    ) {
+        storage.testCertificates.map { it.identifier to it }.toMap().also {
+            Timber.tag(TAG).v("Restored TestCertificate data: %s", it)
+        }
+    }
+
+    val certificates: Flow<Set<TestCertificateContainer>> = internalData.data.map { it.values.toSet() }
+
+    init {
+        internalData.data
+            .onStart { Timber.tag(TAG).d("Observing TestCertificateContainer data.") }
+            .onEach {
+                Timber.tag(TAG).v("TestCertificateContainer data changed: %s", it)
+                storage.testCertificates = it.values.toSet()
+            }
+            .catch {
+                it.reportProblem(TAG, "Failed to snapshot TestCertificateContainer data to storage.")
+                throw it
+            }
+            .launchIn(appScope + dispatcherProvider.Default)
+    }
+
+    suspend fun requestCertificate(test: CoronaTest): TestCertificateContainer {
+        Timber.tag(TAG).d("createDccForTest(test.identifier=%s)", test.identifier)
+
+        val newData = internalData.updateBlocking {
+            if (values.any { it.registrationToken == test.registrationToken }) {
+                Timber.tag(TAG).e("Certificate entry already exists for %s", test.identifier)
+                throw IllegalArgumentException("A certificate was already created for this ${test.identifier}")
+            }
+
+            // TODO do we need additional validation here? last chance to abort?
+
+            val identifier = UUID.randomUUID().toString()
+
+            val certificate = when (test.type) {
+                CoronaTest.Type.PCR -> PCRCertificateContainer(
+                    identifier = identifier,
+                    registeredAt = test.registeredAt,
+                    registrationToken = test.registrationToken,
+                )
+                CoronaTest.Type.RAPID_ANTIGEN -> RACertificateContainer(
+                    identifier = identifier,
+                    registeredAt = test.registeredAt,
+                    registrationToken = test.registrationToken,
+                )
+            }
+            Timber.tag(TAG).d("Adding test certificate entry: %s", certificate)
+            mutate { this[certificate.identifier] = certificate }
+        }
+
+        return newData.values.single { it.registrationToken == test.registrationToken }
+    }
+
+    data class RefreshResult(
+        val certificate: TestCertificateContainer,
+        val error: Exception? = null,
+    )
+
+    suspend fun refresh(identifier: TestCertificateIdentifier? = null): Set<RefreshResult> {
+        Timber.tag(TAG).d("refresh(identifier=%s)", identifier)
+
+        val refreshResults = mutableSetOf<RefreshResult>()
+
+        val workedOnIds = mutableSetOf<TestCertificateIdentifier>()
+
+        internalData.updateBlocking {
+            val toRefresh = values
+                .filter { it.identifier == identifier || identifier == null }
+                .filter { it.isPending }
```

Comment:
can you merge the two `filters ` in one ?

Reply:
Would prefer to keep it this way as performance is not an issue here, we will not have many entries.
The alternative seems less readable.

```kotlin
 .filter { (it.identifier == identifier || identifier == null) && it.isCertificateRetrievalPending  }
```

Semantically it's also nice: First we filter for all of those that are target (all or single identifier), then we filter for those that still need refreshing.

If this is a blocker for you, it can be changed though, I don't feel strongly about it.